### DISCUSSION TITLE
Dynamic tags

### DISCUSSION THREAD

**Discussion by coopbri** [storybookjs/storybook#24192] [Help]
Related: https://github.com/storybookjs/storybook/issues/21761

Is there any way to set tags dynamically, to provide type safety? I am unable to set tags dynamically due to "performance purposes" (quote from linked issue above) regarding static analyzability.

```ts
const meta = {
  tags: ["tag"], // works ✅
};
```

The below examples result in `Error: CSF: Expected tag to be string literal`

```ts
enum Tags {
  Demo = "tag",
}

const meta = {
  tags: [Tags.Demo], // does not work ❌
};
```

```ts
const tags = {
  demo: "tag",
};

const meta = {
  tags: [tags.demo], // does not work ❌
};
```

Even the following does not work:

```ts
tags: ["tag"] as Tags[];
```

My understanding is that there should be no performance issues with type annotations, as in the above example, since that is simply a hint to the TS compiler.

**Comment by connorjs** [comment]
EDIT: DOES NOT SOLVE DISCUSSED PROBLEM

Quick answer: depending on your goals with type safety, you could use the "constants object" approach rather than an (typescript) enum.

const Tags = {
  FOO: "foo",
  BAR: "bar",
  // etc
}

But +1 to another use case to learn more about static analysis needs that prohibit this usage

**Reply by coopbri** [reply]
This appears to be the same as my second example, unless I am misunderstanding you. Have you gotten this to work?

**Reply by connorjs** [reply]
You're right. Updated my comment

**Reply by coopbri** [reply]
Ahh, thanks for your input in any event!

**Comment by shilman** [comment]
On startup we're indexing all the stories in your Storybook and creating a file called `index.json` that contains every story and some metadata around it. If all the metadata definitions are constrained to be defined in the story file itself, it makes the job of that analysis easier and faster.

We might extend the analysis in the future to support imported constructs in the future. But to start with we're keeping it simple and are focusing our efforts elsewhere.

**Reply by shilman** [reply]
For everybody who downvoted, if storybook tags were typesafe using union types, how much of the problem would that solve for you?

**Reply by noveogroup-amorgunov** [reply]
Hello! We use tags for screenshot testing over the storybook. So every story have the follow line:

```ts
const meta: Meta<typeof X> = {
  // ...
  tags: ['{"wait":5000}', '[{"width":"600"}]', '[{"width":"1024"}]'],
}
```

If I try to move this default tags to common var like DEFAULT_TAGS or any tag to separate variable, I get the error:

```ts
const meta: Meta<typeof X> = {
  // ...
  tags: DEFAULT_TAGS, // error
  tags: generateTags({ wait: 5000, isMobile: true, isDesktop: false }), // error
}
```

So it's a problem when we have hundreds of files and we can't handle the automatic generation of tags

**Reply by shilman** [reply]
@noveogroup-amorgunov why don't you use `parameters` for that? You can put whatever you want in there.

https://storybook.js.org/docs/writing-stories/parameters

**Comment by dutscher** [comment]
> EDIT: DOES NOT SOLVE DISCUSSED PROBLEM
> 
> Quick answer: depending on your goals with type safety, you could use the "constants object" approach rather than an (typescript) enum.
> 
> const Tags = { FOO: "foo", BAR: "bar", // etc }
> 
> But +1 to another use case to learn more about static analysis needs that prohibit this usage

I'm on the same opinion as you. The tags should be typesafe and not stringliterals. when we use typescript then this should go.
For now we have no other choice to use stringsLiterals because the csf parser only except strings in an array: https://github.com/storybookjs/storybook/blob/e0c5ff46bbf3ef249f68819e8fc568e9288ecf9d/code/lib/csf-tools/src/CsfFile.ts#L40
I got the point that the babelParser should be fast and the enums arent availble in raw file.
Maybe some day we got this tiny helpful feature? And a docs for the sidebar filters function :D

cheers

**Comment by karol-bujacek** [comment]
Hello,

Do you have any update on this topic?

I have `Error: CSF: Expected tag to be string literal` in Storybook 9.1.16 for vue3-vite. The reason is that I have tried enums for tags.
